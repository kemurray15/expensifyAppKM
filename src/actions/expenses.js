import uuid from 'uuid';
import database from '../firebase/firebase'

/* current flow before integrating firebase */
// component calls action generator
// action generator returns an object
// component dispatches an object
// the redux store changes

/* future flow before integrating firebase */
// component calls action generator
// action generator returns a function
// component dispatches a function
// the function runs

//ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0
        } = expenseData;
        
        const expense = { description, note, amount, createdAt };

        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }))
        })
    };
};

//REMOVE_EXPENSE
export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = ({id} = {}) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove().then(() => { 
            dispatch(removeExpense({id}));
        });
    };
};

//EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

//SET_EXPENSE (get the array back from firebase and update the store and done)

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        
        return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];
            
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            
            console.log(expenses);
            dispatch(setExpenses(expenses));
        });
    };
};


//fetch all expense data once
//parse the data into an array (check firebase.js file)
//dispatch set expenses so that the store actually gets updated