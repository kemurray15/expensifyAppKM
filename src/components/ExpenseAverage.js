import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral'
import selectExpenses from '../selectors/expenses';
import selectExpensesAverage from '../selectors/expenses-average';


export const ExpenseAverage = ( {expenseCount, expenseAverage} ) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses'
    const formattedAverage = numeral(expenseAverage / 100).format('$0,0.00')
    return (
        <div>
            <p>You have {expenseCount} {expenseWord} averaging {formattedAverage} per expense</p>
        </div>
    );
}
    

    
const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    
    return {
        expenseCount: visibleExpenses.length,
        expenseAverage: selectExpensesAverage(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpenseAverage);