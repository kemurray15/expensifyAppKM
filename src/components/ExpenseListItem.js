//export stateless functional component
//description, amount, createdAt value

import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ id, description, amount, createdAt, index}) => (
    <div className="Expense">
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>
            {amount} - {createdAt}       
        </p>
    </div>
);

export default ExpenseListItem;
