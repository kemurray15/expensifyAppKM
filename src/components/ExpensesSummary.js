import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral'
import selectExpenses from '../selectors/expenses';
import expensesTotal from '../selectors/expenses-total';


export const ExpensesSummary = ( {expenseCount, expenseTotal} ) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses'
    const formattedAmount = numeral(expenseTotal / 100).format('$0,0.00')
    return (
        <div>
            <p>Viewing {expenseCount} {expenseWord} totalling {formattedAmount}</p>
        </div>
    );
}
    

    
const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    
    return {
        expenseCount: visibleExpenses.length,
        expenseTotal: expensesTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);