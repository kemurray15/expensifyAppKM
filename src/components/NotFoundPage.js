import React from 'react';
import {Link} from 'react-router-dom';

const NotFoundPage = () => (
    <div>
        {/* We don't want to do the below because that href will point us back to the server
        rather than our javascript*/}
        {/* 404 - <a href="/"> Go home</a> */}
        404 - <Link to="/">Go home</Link>
    </div>
);

export default NotFoundPage;