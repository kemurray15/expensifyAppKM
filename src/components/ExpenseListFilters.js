import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from '../actions/filters';
  
export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };

    //even though the invocation of this method below passes no input, it sounds like
    //this function automatically takes in an object made available by React and we can
    //destructure it to reference the startDate and endDate even though we haven't had
    //to write any code that is making them available

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    //same thing as above - this founction automatically gets passed in the calendarFocused
    // input.  We could call it whatever we want here, its just easiest to use the exact name
    // of our state variable
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }

    onSortChange = (e) => {
        if (e.target.value === 'date') {
            this.props.sortByDate();
        } else if (e.target.value === 'amount') {
            this.props.sortByAmount();
        }
    }

    render() {
        return (
            <div>
                <input 
                    type="text" 
                    value={this.props.filters.text}
                    onChange={this.onTextChange} 
                />
                <select 
                value={this.props.filters.sortBy} 
                onChange={this.onSortChange}        
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={(day) => false}
                    showClearDates={true}
                />
            </div>
        );
    }
};


const mapStateToProps = (state) => ({
    filters: state.filters
});

//allows us to setup property functions that call dispatch.  If we were calling dispatch directly it would be hard or impossible to 
//use jest spies.
const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
