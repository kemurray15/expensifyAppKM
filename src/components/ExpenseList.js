import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

//This component has no connection to the state.  It requires
//another component to render it and pass in any necessary state as props
export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ) : (
                props.expenses.map((expense) => (
                    <ExpenseListItem 
                        key={expense.id}
                        {...expense} 
                    />        
                ))
            )
        }    
    </div>
);


//This is basically saying create a HOC called ConnectedExpenseList.  Connect it
// to the store (though I don't think that's happened yet), and also
//define some other properties that will be passed in to the final rendered
//component, in this case ExpenseList.  Then ExpenseList above can
//reference it's props. sounds like we could also have it return some
//state variables which would then render ExpenseList with some state values

//This function sets up the mapping of state variables to the prop variables
//that will be passed to the rendered child component.  It's just saying
//there is a state.expenses object that I want to just call expenses when
//I pass it in as component props.
const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

//#1This effectively says render ExpenseList with the props mapped from state
//assigned in the above mapStateToProps
//#2When you connect a component to a redux store (I'm still not sure how it
//is connecting to my store) it subscribes in realtime.  So any changes to the
//state will cause the connected components to rerender.
//#2a We also don't need any store.subscribe declarations b/c connect
//is essentially subscribing for us, i.e. saying hey state changed so let me
//remap the new state to props and rerender the component
//#3Interesting that what's exported from this file is not the ExpenseList 
//JSX function alone defined here.  It's the ExpenseList component
//bound to some portion of state renamed for props via the mapStateToProps function
//4The last missing point is {Provider}.  B/c that is the parent component,
//it makes store accessible to all subcomponents.  That is why the
//mapStateToProps function has access to state and I couldn't understand
//how it was getting it.  When this component is a child of Provider
//it will have implicit access to state.  That is also why the mapStateToProps
//is so important - b/c w/o that we would just be giving ExpenseList all of
//the store and that is wasteful.  This way we need only give ExpenseList
//those parts of store that it needs
export default connect(mapStateToProps)(ExpenseList);