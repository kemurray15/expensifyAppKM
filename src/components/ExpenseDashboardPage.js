import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';
import ExpenseAverage from './ExpenseAverage'

const ExpenseDashboardPage = () => (
    <div>
        <ExpenseAverage />
        <ExpensesSummary />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;