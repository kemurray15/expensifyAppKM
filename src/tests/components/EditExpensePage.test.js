import { EditExpensePage } from '../../components/EditExpensePage'
import { shallow } from 'enzyme'
import React from 'react'
import expenses from '../fixtures/expenses'

let editExpense, removeExpense, history, wrapper

beforeEach( () => {
    editExpense = jest.fn();
    removeExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(
        <EditExpensePage 
            editExpense={editExpense} 
            removeExpense={removeExpense} 
            history={history}
            expense={expenses[0]} 
        />
    )
})

//should render EditExpensePage w/ snapshot
test('should render EditExpensePage', () => {
    expect(wrapper).toMatchSnapshot()
})

//should handle editExpense action with spies
test('should handle onSubmit', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0])
    expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0])
    expect(history.push).toHaveBeenLastCalledWith('/');
    
})

// //should handle removeExpense action with spies
test('should handle onClick', () => {
    wrapper.find('button').simulate('click')
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(removeExpense).toHaveBeenLastCalledWith({id: expenses[0].id})
})





