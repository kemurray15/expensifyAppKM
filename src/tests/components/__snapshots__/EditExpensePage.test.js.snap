// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 300,
        "createdAt": 0,
        "description": "bagel",
        "id": "1",
        "note": "everything",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
